### ISSUE TITLE
RevisionGrid: tooltip with multi-line body is not shown if the multi-line indicator column is hidden

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#5036]
**Do you want to request a *feature* or report a *bug*?**

bug/incompleteness in PR #5017 for issue #4990
I'll solve it myself after Drew Noakes will have finished the works on the RevisionGrid.

**What is the current behavior?**

no tooltip with multi-line body in MessageDataGridViewColumn if IsMessageMultilineDataGridViewColumn is hidden, i.e. AppSettings.ShowIndicatorForMultilineMessage == false.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

switch off "Show indicator for multiline message" in the main menu "View"
hover mouse in message column over a commit with a multi-line message

**What is the expected behavior?**

tooltip with complete commit message

**Environment you encounter the issue:**

- GitExtensions version: master
- GIT version: (2.17.1.2)
- OS version: (Windows 7)

**Did this work in previous version of GitExtensions (which)?**

no, new feature

**Comment by RussKie** [comment]
Not sure which way @drewnoakes chooses to refactor the grid, but I would
extract all logic to a renderer, pass it commit data and get it to render
whichever way the user requested.

**Comment by gerhardol** [comment]
The full log message is only retrieved if MultiLine message is retrieved, for performance reasons, this is a quite heavy operation.

In this case, the message should be fetched explicitly if used in the tooltip.

**Comment by drewnoakes** [comment]
As @gerhardol correctly said the `Body` property is set to `null` shortly after the `GitRevision` is created, with the intention of reducing the amount of retained heap memory. On large repos, this can really add up. Currently we only retrieve the full body to support client-side filtering of the message body. I would like to see if we can push such filtering to git operation as well, in which case we wouldn't retried the body at any point during the full load of all revisions.

For now there is no quick and sensible solution to this issue. I know that tooltips on the indicators was probably a big part of why you made #5017. There's still a lot of value there in the support for tooltips of truncated text. But yes, as it stands the inconsistency this bug exposes is not great.

Can we please hold off on addressing this until revision grid refactoring lands? I'm working on it now so it shouldn't take too much longer.

Another issue with #5017 is that moving your mouse over the revision graph fires off git.exe processes in the background. Again, I would like to revisit this once the grid refactor stuff is in place.

> Not sure which way @drewnoakes chooses to refactor the grid

I've been looking at pulling out column provider classes. The grid will delegate/orchestrate between them. I think that should work easily enough.

A few other ideas for the refactor too, though mostly I've been grouping related chunks of code together to help analyse what can be pulled out. This process gives a good result, but wreaks havoc in the diff so if something good comes from it, there'll be another heavy PR involved. As always I'll keep the commits short and clear to aid review.

**Comment by vbjay** [comment]
@drewnoakes  Along with the other refactors of the grid, what do you think of https://docs.microsoft.com/en-us/dotnet/framework/winforms/controls/implementing-virtual-mode-wf-datagridview-control?

**Comment by mstv** [comment]
> Can we please hold off on addressing this until revision grid refactoring lands?

Yes, of course. (I mentioned this at the very beginning of this issue.)

> Another issue with #5017 is that moving your mouse over the revision graph fires off git.exe processes in the background.

Maybe the branches shouldn't be displayed in the tooltip — at least not that way. I tried to ask for comments on this in #4990 (see item 1 of Expected Behavior).
As an alternative or final goal: I've played around following the DvcsGraph.Junctions to find the branch a commit was committed to. (I know this cannot be done for all cases because Git does not want to store this information.) This would work without firing off git.exe.

**Comment by RussKie** [comment]
> A few other ideas for the refactor too, though mostly I've been grouping related chunks of code together to help analyse what can be pulled out. This process gives a good result, but wreaks havoc in the diff so if something good comes from it, there'll be another heavy PR involved. As always I'll keep the commits short and clear to aid review.

I was thinking along these lines - #4376 (439212acdcb8e7e45c3b0c7b603346e6a02dfe8c). Perhaps we can marry up both approaches.

**Comment by RussKie** [comment]
> I've been looking at pulling out column provider classes. The grid will delegate/orchestrate between them. I think that should work easily enough.

I was thinking along these lines - #4376 (439212a). Perhaps we can marry up both approaches.

**Comment by RussKie** [comment]
My idea was to abstract layout render, so we could remove and add layouts
as/if necessary.
If we want to re-add layouts later we probably want to plan it now

**Comment by drewnoakes** [comment]
@mstv this should be fixed on the `v3-ux` branch. If you get a chance to check it out, please do.

Note that the fix involves a compromise where we only retain the full commit message for commits authored in the last six months. Commits earlier than that include a message explaining the limitation in the tooltip.

**Comment by mstv** [comment]
> Note that the fix involves a compromise where we only retain the full commit message for commits authored in the last six months. Commits earlier than that include a message explaining the limitation in the tooltip.

I like the compromise. The UX is very clear.

This issue has already become obsolete in the `v3-ux` branch: The data retrieval has been optimized and the option to hide the multi-line indicator has been removed.
I want to keep this issue open until the `v3-ux` branch will be finished in order to check it again.
